server/asset/dcr/dcr.go

```diff
@@ -921,14 +924,19 @@ func (dcr *Backend) getUnspentTxOut(ctx context.Context, txHash *chainhash.Hash,
 // Get information for an unspent transaction output, plus the verbose
 // transaction.
 func (dcr *Backend) getTxOutInfo(ctx context.Context, txHash *chainhash.Hash, vout uint32) (*chainjson.GetTxOutResult, *chainjson.TxRawResult, []byte, error) {
-	txOut, pkScript, err := dcr.getUnspentTxOut(ctx, txHash, vout)
-	if err != nil {
-		return nil, nil, nil, err
-	}
 	verboseTx, err := dcr.node.GetRawTransactionVerbose(ctx, txHash)
 	if err != nil {
 		return nil, nil, nil, fmt.Errorf("GetRawTransactionVerbose for txid %s: %w", txHash, translateRPCCancelErr(err))
 	}
+	msgTx, err := msgTxFromHex(verboseTx.Hex)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to decode MsgTx from hex for transaction %s: %w", txHash, err)
+	}
+	txTree := stake.DetermineTxType(msgTx, msgTx.Version == wire.TxVersionTreasury)
```

Comment:
Kind of makes me wonder why `DetermineTxType` doesn't just check the version itself.

Reply:
Right. Gonna ping @davecgh for this.

Reply:
It depends on the result of the treasury vote.  This only works here because dcrd is actually enforcing the condition is true, which it couldn't do if it were relying on the very condition it has to enforce.

Reply:
As an aside, this code does not look accurate.  `DetermineTxType` doesn't return the `txTree`, it returns the transaction type so it's passing the wrong thing to getUnspentTxOut, which is why you're having to cast it too.

```Go
	txType := stake.DetermineTxType(tx.MsgTx(), isTreasuryEnabled)
	tree := wire.TxTreeRegular
	if txType != stake.TxTypeRegular {
		tree = wire.TxTreeStake
	}
	txOut, pkScript, err := dcr.getUnspentTxOut(ctx, txHash, vout, tree)
```

Reply:
Also, I didn't look at the surrounding code, but that `msgTx.Version == wire.TxVersionTreasury` is only guaranteed to be accurate for the first transaction in the stake tree.  You can't just look at any old transaction's version and conclude that.

Reply:
Does said dcrd enforcement apply only to mined transactions or can we take this shortcut for transactions in mempool too?
Also, should this be `msgTx.Version >= wire.TxVersionTreasury` rather than just `==`.

Reply:
Oh, I see you just said that this shortcut isn't even valid for arbitrary txns anyway, only the first in the stake tree.

Well, can we always say `true` then?  If it is actually `false`, then there won't be any treasury txn types or new vote versions since dcrd wouldn't allow them, right?

Reply:
I see that bad code now @davecgh. I mean the function says determine **txtype** not tree!

Re the `msgTx.Version == wire.TxVersionTreasury` check guarantee, it came up in chat a while ago (about the time the `tree` param was added to the `gettxout` method and it was suggested that this check is sufficient to tell if treasury is enabled. If it's not sufficient and we do not have access to the full block to pull the first stake tx, is there a better alternative? Or simply just `true` as is done in a couple other call sites I've seen?

Reply:
> 
> 
> Oh, I see you just said that this shortcut isn't even valid for arbitrary txns anyway, only the first in the stake tree.
> 
> Well, can we always say `true` then? If it is actually `false`, then there won't be any treasury txns or new vote versions since dcrd wouldn't allow them, right? So in this case `true` would just be inefficient, but it wouldn't falsely identify certain regular transactions as stake.

It depends on exactly how it's being used.  The only things that changed about the old transactions are:

1) Regular tree coinbase transactions
2) Stake tree votes

All of the new treasury transactions will obviously be prevented before the vote passes by dcrd, so you really wouldn't have any issue there in terms of just using `true`.

However, you will have an issue with it detecting current coinbases and votes if you pass it `true` prior to treasury activation because their format changed.

Reply:
It's not the prettiest, but a potential solution is to just check both and if **neither** come up as a stake transaction, you'll be fine, because dcrd will prevent the wrong ones from making it into a block accordingly.

e.g. something like:

```Go
func determineTxTree(msgTx *wire.MsgTx) int8 {
	// Try with treasury disabled first.
	txType := stake.DetermineTxType(msgTx, false)
	if txType != stake.TxTypeRegular {
		return wire.TxTreeStake
	}

	// Try with treasury enabled.
	txType := stake.DetermineTxType(msgTx, true)
	if txType != stake.TxTypeRegular {
		return wire.TxTreeStake
	}

	return wire.TxTreeRegular
}

...

tree := determineTxTree(msgTx)
txOut, pkScript, err := dcr.getUnspentTxOut(ctx, txHash, vout, tree)
```

Reply:
Thanks for bearing with me @davecgh. I want to get clear on this for dcrdata as well.

> All of the new treasury transactions will obviously be prevented before the vote passes by dcrd, so you really wouldn't have any issue there in terms of just using `true`.
> 
> However, you will have an issue with it detecting current coinbases and votes if you pass it `true` prior to treasury activation because their format changed.

For coinbases, I have noted that `blockchain/standalone.IsCoinBaseTx` have an issue if we say `true` prematurely:

https://github.com/decred/dcrd/blob/0a8c52a100670babcf1f097c7415cf4ef8ceb1e2/blockchain/standalone/tx.go#L38-L42

However, for `stake.DetermineTxType`, it appears that a coinbase would just remain in the default case (`TxTypeRegular`) after (still) matching none of the stake types regardless of `isTreasuryEnabled`.

For votes, that's a question of what [`stake.CheckSSGenVotes`](https://github.com/decred/dcrd/blob/0a8c52a100670babcf1f097c7415cf4ef8ceb1e2/blockchain/stake/staketx.go#L868) will do.  In this case it seems that saying `true` incorrectly/prematurely will also not be a problem, but I'm less certain about this:

https://github.com/decred/dcrd/blob/0a8c52a100670babcf1f097c7415cf4ef8ceb1e2/blockchain/stake/staketx.go#L999-L1010

Saying `true` with `stake.CheckSSGenVotes` appears to have no effect for old votes whose last output is not a nulldata output.

Reply:
Yeah, I think you can be a little loose in dcrdata with the vote check and just use `true` for that since dcrd isn't going to let bad votes in and it appears old votes will still be identified when given `true`.  I actually thought that wasn't the case, but after looking at it again, I see that it will because the final treasury spend bits are allowed to be omitted.

So, even though it's not technically accurate. since dcrdata is only ever looking at votes in the mempool (which comes from dcrd via RPC, right?) and blocks which have been validated by dcrd, dcrdata should be protected from misidentifying them.

For the coinbase case, since they aren't allowed in the mempool, you could get away with only checking with and without treasury enabled as I suggested above for the first tx of the regular tree and that would also always be correct since dcrd will prevent the wrong types from making it into a block.